### DISCUSSION TITLE
How can I optimize a MenuButton+Menu when rendered in 500 table rows?

### DISCUSSION THREAD

**Discussion by jasikpark** [ariakit/ariakit#3118] [General]
I have a table that can have a large number of rows, at the end of each row is a Menu:

```tsx
return (
    <>
      <MenuButton
        store={menu}
        className={`btn icon-btn ${styles.MeatballMenu_button}`}
        aria-label={ariaLabel}
        disabled={disabled}
      >
        <Meatballs />
      </MenuButton>

      <Menu store={menu} className={cn('menu-list', className)} portal={false}>
        {children}
      </Menu>
    </>
  );
```

I have root-caused my lag/loading issues down to the `Menu`; If I comment out

```tsx
 <Menu store={menu} className={cn('menu-list', className)} portal={false}>
        {children}
      </Menu>
```

then I end up w/ a relatively lag-free list page.

Currently trying to learn how to profile and read flamegraphs 😓 

`<Menu />` source: https://github.com/ariakit/ariakit/blob/main/packages/ariakit-react-core/src/menu/menu.tsx

**Comment by diegohaz** [comment]
Try [`unmountOnHide`](https://ariakit.org/reference/menu#unmountonhide):
```jsx
<Menu
  store={menu}
  portal={false}
  unmountOnHide
  className={cn('menu-list', className)}
>
  {children}
</Menu>
```

**Reply by jasikpark** [reply]
I need to better test things, but that does not seem to make a noticeable difference for me

In this instance `{children}` is:

```tsx
        <MenuItem render={<Link to={encodeURIComponent(tagName)} />}>Edit</MenuItem>
        <MenuItem onClick={() => setTagBeingDeletedName(tagName)}>Delete</MenuItem>
```

(thought entirely removing the children of `<Menu>` doesn't seem to change things, unlike removing the actual `<Menu>` component)

**Reply by diegohaz** [reply]
Are you using the latest version?

**Reply by jasikpark** [reply]
I am once you asked me to use that new prop - I'm using the lib w/ Typescript 👍 

Is there a perf advantage to `useMenuStore()` v.s. `<MenuProvider>`?

**Reply by jasikpark** [reply]
Going to start w/ 


```diff
diff --git a/src/components/patterns/menus/meatball-menu/MeatballMenu.tsx b/src/components/patterns/menus/meatball-menu/MeatballMenu.tsx
index 7916dd31..dc711bde 100644
--- a/src/components/patterns/menus/meatball-menu/MeatballMenu.tsx
+++ b/src/components/patterns/menus/meatball-menu/MeatballMenu.tsx
@@ -1,5 +1,4 @@
 import cn from 'classnames';
-import { useEffect } from 'react';
 import { Menu, MenuButton, useMenuStore } from '@components/patterns/menus/menu';
 import styles from './MeatballMenu.module.css';
 
@@ -7,31 +6,11 @@ type Props = {
   children: React.ReactNode;
   className?: string;
   'aria-label'?: string;
-  /** Make sure that this is memoized, if it changes each render, it will cause an infinite loop */
-  onOpen?: () => void;
-  /** Make sure that this is memoized, if it changes each render, it will cause an infinite loop */
-  onClose?: () => void;
   disabled?: boolean;
 };
 
-export function MeatballMenu({
-  children,
-  onOpen,
-  onClose,
-  className,
-  disabled,
-  'aria-label': ariaLabel = 'More',
-}: Props) {
+export function MeatballMenu({ children, className, disabled, 'aria-label': ariaLabel = 'More' }: Props) {
   const menu = useMenuStore();
-  const isExpanded = menu.useState('open');
-
-  useEffect(() => {
-    if (isExpanded) {
-      onOpen?.();
-    } else {
-      onClose?.();
-    }
-  }, [onOpen, onClose, isExpanded]);
 
   return (
     <>
@@ -43,8 +22,7 @@ export function MeatballMenu({
       >
         <Meatballs />
       </MenuButton>
-
-      <Menu store={menu} className={cn('menu-list', className)} portal={true}>
+      <Menu store={menu} className={cn('menu-list', className)} portal={false} unmountOnHide={true}>
         {children}
       </Menu>
     </>
```

Seems like a good simplification, slightly improves perf.

**Reply by diegohaz** [reply]
> Is there a perf advantage to `useMenuStore()` v.s. `<MenuProvider>`?

There shouldn't be a noticeable difference. `MenuProvider` is just a component that wraps `useMenuStore`.

You can try to conditionally render `<Menu>` when the menu is open:

```jsx
const mounted = menu.useState("mounted");
mounted && <Menu ...>
```

However, that's exactly what `unmountOnHide` does under the hood. So I'm not sure if it will help.

Alternatively, you can try to render a single `<Menu>` for multiple `<MenuButton>`, which is now supported:

```jsx
const menu = useMenuStore();
const tagName = menu.useState((state) => state.disclosureElement?.dataset.tag);

<MenuButton store={menu} data-tag="a" />
<MenuButton store={menu} data-tag="b" />
<MenuButton store={menu} data-tag="c" />

<Menu store={menu}>
  <MenuItem render={<Link to={encodeURIComponent(tagName)} />}>Edit</MenuItem>
  <MenuItem onClick={() => setTagBeingDeletedName(tagName)}>Delete</MenuItem>
</Menu>
```

I haven't tested this code, but it should work, in theory.

**Reply by jasikpark** [reply]
I guess I could probably write up a codesandbox to mess around with at some point...

The main code is a tanstack table w/ the Menu in one of the columns, with 500 rows with unique, simple component props.

**Reply by diegohaz** [reply]
You can also implement your `onOpen` and `onClose` props using [state setters](https://ariakit.org/guide/component-stores#state-setters):

```jsx
useMenuStore({
  setOpen(open) {
    if (open) {
      onOpen?.();
    } else {
      onClose?.();
    }
  }
});
```

Those props don't need to be memoized in this case.

**Reply by jasikpark** [reply]
Cool! I ended up just removing the onOpen/onClose since we never used them in the app 🤷

**Reply by diegohaz** [reply]
Did that help with the perf issue? Otherwise, if you put together a CodeSandbox/StackBlitz I can take a better look.

**Reply by jasikpark** [reply]
Removing the `useEffect` definitely helped slightly, I'm going to drop this issue for the moment, but I have a perf issue in my project to revisit this problem / the topic later
